### PR TITLE
feat(feature-activation): implement caching mechanism

### PR DESCRIPTION
**Pull request by glevco**
### Acceptance Criteria

- Implement `TransactionMetadata.feature_states` attribute
- Implement `Block.get_feature_state()` and `Block.update_feature_state()` methods
- Implement feature state caching in `FeatureService` using block metadata

Resolves https://github.com/HathorNetwork/hathor-core/issues/630

### CODE REVIEW

File: hathor/simulator/simulator.py [hunk 1]
```diff
@@ -149,7 +150,8 @@ def create_peer(
         enable_sync_v2: bool = True,
         soft_voided_tx_ids: Optional[Set[bytes]] = None,
         full_verification: bool = True,
-        event_ws_factory: Optional[EventWebsocketFactory] = None
+        event_ws_factory: Optional[EventWebsocketFactory] = None,
+        tx_storage: Optional[TransactionStorage] = None
```
Comment (msbrogli) [inline hathor/simulator/simulator.py:154]: Why do you need to set `tx_storage`? I was considering to refactor all tests to remove `Builder.set_storage()`. Can't you use `Builder.use_memory()` or `Builder.use_rocksdb()` instead?

**Review by msbrogli — APPROVED** [review]

**Review by jansegre — APPROVED** [review]

File: hathor/simulator/simulator.py (hunk 1 above)
Comment (glevco) [reply hathor/simulator/simulator.py]: This has been removed after https://github.com/HathorNetwork/hathor-core/pull/653 was merged to master and I rebased it here

Comment (msbrogli) [reply hathor/simulator/simulator.py]: Ok. Thanks!

**Review by jansegre — DISMISSED** [review]

File: tests/feature_activation/test_feature_simulation.py [hunk 2]
```diff
@@ -41,46 +47,57 @@
...
-            feature_settings=_FEATURE_SETTINGS,
-            feature_service=self.feature_service,
-            tx_storage=manager.tx_storage
-        )
-        self.web_client = StubSite(feature_resource)
-
-        self.miner = self.simulator.create_miner(manager, hashpower=1e6)
-        self.miner.start()
+class BaseFeatureSimulationTest(SimulatorTestCase):
+    seed_config = 5604003716498505253
```
Comment (msbrogli) [inline tests/feature_activation/test_feature_simulation.py:51]: Why are you using a fixed seed?

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: tests/feature_activation/test_feature_simulation.py (hunk 2 above)
Comment (glevco) [reply tests/feature_activation/test_feature_simulation.py]: Sorry, this was a rebase artifact. Removed it again in https://github.com/HathorNetwork/hathor-core/pull/645/commits/c050c149f315032b9d068f45113d71d31db4d083

Comment (msbrogli) [reply tests/feature_activation/test_feature_simulation.py]: Ok!

**Review by msbrogli — APPROVED** [review]

**Review by jansegre — APPROVED** [review]

**Pull request merged**